### PR TITLE
Add feature that pushes a unique version of containers into a cluster and requires a specific version of concent and golem

### LINKED ISSUE #134: Unique container versions
Currently our containers are tagged using the output of `git describe --always` which gives output like `v0.5.3-rc3-2-g3d1eb9a`. Those tags are not pushed to the Docker registry though and also they do not uniquely identify a particular build since they depend only on concent-deployment and specific Concent and Golem versions are not hard-coded in that repository.

- Make [kubernetes/Makefile#L57-L60](https://github.com/golemfactory/concent-deployment/blob/master/kubernetes/Makefile#L57-L60) construct the tag the same way [containers/Makefile#L49-L52](https://github.com/golemfactory/concent-deployment/blob/master/containers/Makefile#L49-L52) does.
    - This will embed the actual version in all scripts and Kubernetes configs.
- Add golem and concent versions to `containers/versions.yml` so that every version of concent-deployment is always associated with specific versions of those repositories.
    - The version should always be either a git tag or a git commit.
- Move cloning golem and concent repositories from the build playbook to the `Makefile`. Put them somewhere inside the `build/` directory and adjust all paths in the `Makefile` accordingly.
    - They should still be cloned from the local clone that we have in `/var/`.
    - You can also remove fetching concent-deployment-values completely (unless we're using it when building `kubernetes/` but I think we're always using the version uploaded from the local machine).

### PR DESCRIPTION
**Pull request by bartoszbetka**
Resolves #134

### CODE REVIEW

File: concent-builder/repositories.yml [hunk 1]
```diff
@@ -3,7 +3,5 @@ repositories:
     url: https://github.com/golemfactory/concent.git
   concent-deployment:
     url: https://github.com/golemfactory/concent-deployment.git
-  concent-deployment-values:
-    url: https://github.com/golemfactory/concent-deployment-values.git
```
Comment (cameel) [inline concent-builder/repositories.yml]: This should be in a separate commit.

File: containers/Makefile [hunk 2]
```diff
@@ -36,15 +36,8 @@ VERIFIER_SOURCE :=                           \
...
-    CONCENT_DIR:= $(abspath ../../concent)
-endif
-$(info CONCENT_DIR=$(CONCENT_DIR))
-
-ifndef GOLEM_DIR
-    GOLEM_DIR:= $(abspath ../../golem)
-endif
-$(info GOLEM_DIR=$(GOLEM_DIR))
+CONCENT_DIR:= build/repositories/concent/
+GOLEM_DIR:= build/repositories/golem/
```
Comment (cameel) [inline containers/Makefile]: Please align and fix whitespace here.

File: containers/versions.yml [hunk 3]
```diff
@@ -1,6 +1,8 @@
 debian_version:               "9.4"
 docker_version:               "17.05.0~ce-0~debian-stretch"
 geth_version:                 "v1.8.2"
+golem_version:                "0.13.0"
+concent_version:              "v0.6.0-39-gb882ecc"
```
Comment (cameel) [inline containers/versions.yml]: So it's possible to use the output of `git describe` with `git checkout`? Cool. I did not expect that.

File: containers/package-builder/checkout-concent-version.sh.j2 [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+DIRECTORY=$1
+
+# Check if a directory exists
+if [ -d "$DIRECTORY" ]; then
+    git fetch
+else
+    git clone /var/concent-data/concent/ "$DIRECTORY"
+    cd "$DIRECTORY"
+    git fetch
+    git checkout {{ concent_version }}
```
Comment (cameel) [inline containers/package-builder/checkout-concent-version.sh.j2]: Please make it a parameter of the script.

File: containers/package-builder/checkout-concent-version.sh.j2 [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+DIRECTORY=$1
+
+# Check if a directory exists
+if [ -d "$DIRECTORY" ]; then
+    git fetch
+else
+    git clone /var/concent-data/concent/ "$DIRECTORY"
+    cd "$DIRECTORY"
+    git fetch
```
Comment (cameel) [inline containers/package-builder/checkout-concent-version.sh.j2]: You do `git fetch` in both cases so you could put it after the loop.

Comment (cameel) [inline containers/package-builder/checkout-concent-version.sh.j2]: When running `git fetch`, always specify the remote name. Otherwise the behavior is ambiguous maybe affected by git configuration.

File: containers/package-builder/checkout-concent-version.sh.j2 [hunk 6]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash -e
+
+DIRECTORY=$1
```
Comment (cameel) [inline containers/package-builder/checkout-concent-version.sh.j2]: Use lower case characters when naming shell variables.

File: containers/package-builder/checkout-golem-version.sh.j2 [hunk 7]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash -e
```
Comment (cameel) [inline containers/package-builder/checkout-golem-version.sh.j2]: You could have a single script that works for both golem and concent if you added more parameters.

File: containers/package-builder/checkout-concent-version.sh.j2 [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+# Check if a directory exists
+if [ -d "$DIRECTORY" ]; then
+    git fetch
+else
+    git clone /var/concent-data/concent/ "$DIRECTORY"
```
Comment (cameel) [inline containers/package-builder/checkout-concent-version.sh.j2]: Please make a shallow clone. There's no need to clone the whole repository if we only need the last revision. See the parameters that were used in the Ansible playbook.

File: containers/Makefile [hunk 9]
```diff
@@ -135,14 +128,20 @@ $(VERIFIER_SOURCE_FROM_CONCENT_API): build/verifier/%: build/concent-api/%
...
 
-build/verifier/golem/imgverifier-requirements.txt: $(GOLEM_DIR)/apps/rendering/resources/images/requirements.txt
+build/verifier/golem/imgverifier-requirements.txt: $(GOLEM_DIR)apps/rendering/resources/images/requirements.txt
 	cp "$<" "$@"
 
+$(CONCENT_DIR): build/package-builder/checkout-concent-version.sh /var/concent-data/concent
+	"$<" "$@"
+
+$(GOLEM_DIR): build/package-builder/checkout-golem-version.sh /var/concent-data/golem
+	"$<" "$@"
```
Comment (cameel) [inline containers/Makefile]: Please hard-code the script name here. It's easier to understand that way.

File: containers/Makefile (hunk 2 above)
Comment (cameel) [inline containers/Makefile]: Please remove these variables. They can no longer be changed by the user who runs make. Leaving them here may imply that they can be changed.

File: concent-builder/repositories.yml (hunk 1 above)
Comment (bartoszbetka) [reply concent-builder/repositories.yml]: Done.

File: containers/Makefile (hunk 2 above)
Comment (bartoszbetka) [reply containers/Makefile]: Done.

File: containers/package-builder/checkout-concent-version.sh.j2 (hunk 4 above)
Comment (bartoszbetka) [reply containers/package-builder/checkout-concent-version.sh.j2]: Could you be more specific ? How do you return the template variable at the script input?

File: containers/package-builder/checkout-concent-version.sh.j2 (hunk 5 above)
Comment (bartoszbetka) [reply containers/package-builder/checkout-concent-version.sh.j2]: Done.

Comment (bartoszbetka) [reply containers/package-builder/checkout-concent-version.sh.j2]: Done.

File: containers/package-builder/checkout-concent-version.sh.j2 (hunk 6 above)
Comment (bartoszbetka) [reply containers/package-builder/checkout-concent-version.sh.j2]: Done.

File: containers/package-builder/checkout-concent-version.sh.j2 (hunk 4 above)
Comment (cameel) [reply containers/package-builder/checkout-concent-version.sh.j2]: I mean that instead of this you could have something like
``` bash
version=$2
git checkout $version
```
and call the script like
``` bash
./checkout-concent-version.sh <version>
```

Oh, but I can see the problem now. Version is embedded in `version.yml` and you don't have it directly in the makefile. Well, in that case let's leave it as is.

File: containers/package-builder/checkout-golem-version.sh.j2 (hunk 7 above)
Comment (bartoszbetka) [reply containers/package-builder/checkout-golem-version.sh.j2]: Done.

File: containers/package-builder/checkout-concent-version.sh.j2 (hunk 8 above)
Comment (bartoszbetka) [reply containers/package-builder/checkout-concent-version.sh.j2]: There is no parameter `--depth` in ansible playbook. This not work, because we clone repository locally.

File: containers/Makefile (hunk 9 above)
Comment (bartoszbetka) [reply containers/Makefile]: Done.

File: containers/package-builder/checkout-concent-version.sh.j2 (hunk 4 above)
Comment (bartoszbetka) [reply containers/package-builder/checkout-concent-version.sh.j2]: Ok.

File: containers/package-builder/checkout-repositories-version.sh.j2 [hunk 10]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash -e
+
+directory=$1
+repository=$2
```
Comment (cameel) [inline containers/package-builder/checkout-repositories-version.sh.j2]: Please reverse the order of these parameters. The script is basically a wrapper over `git clone` and the order in clone is `<repository> <directory>`. It will make it easier to remember.

File: containers/package-builder/checkout-repositories-version.sh.j2 [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+        git clone /var/concent-data/golem/ "$directory"
+        cd "$directory"
+        git checkout {{ golem_version }}
+fi
+
+if [[ "$repository" == "concent" ]]; then
+git fetch https://github.com/golemfactory/concent.git
+elif [[ "$repository" == "golem" ]]; then
+git fetch https://github.com/golemfactory/golem.git
+else
```
Comment (cameel) [inline containers/package-builder/checkout-repositories-version.sh.j2]: Please fix indentation here.

File: containers/package-builder/checkout-repositories-version.sh.j2 [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+if [ ! -d "$directory" ] && [[ "$repository" == "golem" ]]; then
+        git clone /var/concent-data/golem/ "$directory"
+        cd "$directory"
+        git checkout {{ golem_version }}
+fi
+
+if [[ "$repository" == "concent" ]]; then
+git fetch https://github.com/golemfactory/concent.git
+elif [[ "$repository" == "golem" ]]; then
+git fetch https://github.com/golemfactory/golem.git
```
Comment (cameel) [inline containers/package-builder/checkout-repositories-version.sh.j2]: Don't fetch from github. You should fetch from the repository in `/var/`.

File: containers/package-builder/checkout-repositories-version.sh.j2 [hunk 13]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash -e
+
+directory=$1
+repository=$2
+
+# Check if a directory exists
+if [ ! -d "$directory" ] && [[ "$repository" == "concent" ]]; then
```
Comment (cameel) [inline containers/package-builder/checkout-repositories-version.sh.j2]: `repository` should be `/var/concent-data/concent/`, not `/concent/`. Similarly for `golem`. This way you won't have to hard-code it for `concent` and `golem`.

File: containers/Makefile [hunk 14]
```diff
@@ -135,14 +125,20 @@ $(VERIFIER_SOURCE_FROM_CONCENT_API): build/verifier/%: build/concent-api/%
...
 
-build/verifier/golem/imgverifier-requirements.txt: $(GOLEM_DIR)/apps/rendering/resources/images/requirements.txt
+build/verifier/golem/imgverifier-requirements.txt: build/repositories/golem/apps/rendering/resources/images/requirements.txt
 	cp "$<" "$@"
 
+build/repositories/concent/: build/package-builder/checkout-repositories-version.sh /var/concent-data/concent
+	build/package-builder/checkout-repositories-version.sh "$@" "concent"
+
+build/repositories/golem/: build/package-builder/checkout-repositories-version.sh /var/concent-data/golem
+	build/package-builder/checkout-repositories-version.sh "$@" "golem"
```
Comment (cameel) [inline containers/Makefile]: Please rename the script to `clone-or-update.sh` or maybe `get-code.sh`.

File: containers/package-builder/checkout-repositories-version.sh.j2 (hunk 10 above)
Comment (bartoszbetka) [reply containers/package-builder/checkout-repositories-version.sh.j2]: Done.

File: containers/package-builder/checkout-repositories-version.sh.j2 (hunk 11 above)
Comment (bartoszbetka) [reply containers/package-builder/checkout-repositories-version.sh.j2]: Done.

File: containers/package-builder/checkout-repositories-version.sh.j2 (hunk 12 above)
Comment (bartoszbetka) [reply containers/package-builder/checkout-repositories-version.sh.j2]: Done.

File: containers/package-builder/checkout-repositories-version.sh.j2 (hunk 13 above)
Comment (bartoszbetka) [reply containers/package-builder/checkout-repositories-version.sh.j2]: Done.

File: containers/package-builder/clone-or-update.sh.j2 [hunk 15]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash -e
+
+repository=$1
+directory=$2
+
+# Check if a directory exists
+if [ ! -d "$directory" ] && [[ "$repository" == "/var/concent-data/concent" ]]; then
+    git clone "$repository" "$directory"
+    cd "$directory"
+    git checkout {{ concent_version }}
```
Comment (cameel) [inline containers/package-builder/clone-or-update.sh.j2]: The code in both `if`s is now identical. You can replace it with a single `if`.

Comment (bartoszbetka) [reply containers/package-builder/clone-or-update.sh.j2]: It's not possible to do that. How would you recognize which template variable use ?

File: containers/package-builder/clone-or-update.sh.j2 [hunk 16]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash -e
+
+repository=$1
+directory=$2
+
+if [[ "$repository" == "concent" ]]; then
+    repository_path=/var/concent-data/concent/
+    repository_version={{ concent_version }}
```
Comment (cameel) [inline containers/package-builder/clone-or-update.sh.j2]: Add quotes around `{{ }}` and it's good.
